### PR TITLE
Remove remaining global `set_default_dtype` calls from tests

### LINKED ISSUE #68972: default dtype is incorrectly set to torch.double when running test_ops.py
Imports into test files should not be setting the default dtype. We should add asserts to the test files after imports that the default dtype is float32. 

@eellison maybe leaking from a jit include? 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #68972

cc @mruberry

### CODE REVIEW

**Comment by kurtamohler** [comment]
I'm having trouble fixing the failing `python test/distributions/test_distributions.py -k test_cdf_icdf_inverse` test in [macos-12-py3-arm64 / test (default, 1, 3, macos-m1-12)](https://github.com/pytorch/pytorch/actions/runs/5884003265/job/15958041824?pr=107246#logs).

I don't see why the input would have changed for that test, since I've wrapped it in a `set_default_dtype(torch.double)` context manager to keep the default dtype the same as it used to be.

Apparently macos jobs don't have ssh set up, so I can't just go in to debug it--unless there's some way to enable ssh for those.

**Comment by lezcano** [comment]
cc @kulinseth

**Review by ezyang — APPROVED** [review]

File: test/distributions/test_distributions.py [hunk 1]
```diff
@@ -103,694 +99,701 @@ def is_all_nan(tensor):
...
-        {'probs': 1.00001},
-    ])
-]
+
+# Register all distributions for generic tests.
+def _get_examples(dtype=None):
+    if dtype is None:
+        dtype = torch.get_default_dtype()
+
+    with set_default_dtype(dtype):
```
Comment (kurtamohler) [inline test/distributions/test_distributions.py:109]: It should not be necessary to have this `dtype` arg if the caller is wrapping the call with a `set_default_dtype()` context manager. But I'm just trying this to see if for some reason this fixes the `test_cdf_icdf_inverse` failure in the [macos-12-py3-arm64 / test (default, 1, 3, macos-m1-12)](https://github.com/pytorch/pytorch/actions/runs/5884003265/job/15958041824?pr=107246#logs) job.

Comment (kurtamohler) [reply test/distributions/test_distributions.py:109]: It didn't work. Removing this.

File: test/distributions/test_distributions.py [hunk 2]
```diff
@@ -103,694 +99,697 @@ def is_all_nan(tensor):
...
+            {'logits': torch.tensor([[0.0, 0.0], [0.0, 0.0]], requires_grad=True)},
+        ]),
+        Example(Pareto, [
+            {
+                'scale': 1.0,
+                'alpha': 1.0
+            },
+            {
+                'scale': (torch.randn(5, 5).abs() + 0.1).requires_grad_(),
+                'alpha': (torch.randn(5, 5).abs() + 0.1).requires_grad_()
```
Comment (kurtamohler) [inline test/distributions/test_distributions.py:380]: I've added these `+ 0.1`'s here to fix the `macos-12-py3-arm64 / test` failure. Earlier, I added some prints to the failing `test_cdf_icdf_inverse` test and looked at the printouts from the CI log. I found out that the samples generated by the Pareto distribution sometimes contain infinity if the values in `alpha` are too close to 0, and then after those infinities go through `cdf` and `icdf`, they turn into `nan`'s, causing comparisons to fail. I copied the exact `scale` and `alpha` values from the failing test into the following snippet:

```python
import torch

for seed in range(100_000):
    torch.manual_seed(seed)
    print(seed)

    param = {
        'scale': torch.tensor([
            [1.7549, 1.1252, 1.1057, 1.2407, 0.8933],
            [0.7661, 0.9723, 0.2869, 1.1382, 0.1948],
            [0.3116, 0.5687, 0.0172, 0.2296, 0.1637],
            [0.6526, 0.8761, 0.0142, 2.0804, 1.3138],
            [0.8522, 2.6757, 0.9873, 0.6828, 1.1830]],
            dtype=torch.double, requires_grad=True),
        'alpha': torch.tensor([
            [0.3219, 2.0156, 0.5705, 1.0555, 0.0485],
            [0.1138, 0.1064, 0.4582, 0.3166, 0.0073],
            [0.3978, 0.3402, 0.4575, 2.7605, 0.3148],
            [0.0442, 1.2770, 1.3061, 1.4474, 0.6836],
            [0.1606, 1.1500, 0.2150, 0.2591, 0.1284]],
            dtype=torch.double, requires_grad=True)
    }

    dist = torch.distributions.Pareto(**param)
    samples = dist.sample(sample_shape=(20,))

    assert not samples.isinf().any()
```

The assert throws an error for me once it gets to `seed=6`, so there is some luck involved--the test evidently used to pass in CI just because we were getting lucky with the random seeds.

If I add `0.1` to `alpha` in the above script, it does not raise an error for any of the 100,000 seeds that the script tries

**Comment by kurtamohler** [comment]
My fix for the `test_cdf_icdf_inverse` failure worked.

Not sure why I'm getting upstream failures when my branch is based off `viable/strict`, but looks like I can finally merge this shortly if no other failures happen

**Comment by kurtamohler** [comment]
I imagine it's possible that some Meta-internal tests could be setting the default dtype, and if so this would probably be reverted if we merge it now, since I added checks for float being the default dtype in the test setup and teardown methods. It may be a good idea to import this and run the internal tests. What do you think, @ezyang ?

**Comment by ezyang** [comment]
FB tests seem basically fine. Merge at will.

**Comment by huydhn** [comment]
@pytorchbot drci

(Please ignore this, I'm testing Dr. CI)

**Comment by kurtamohler** [comment]
Well that's unfortunate. It's not clear to me why it was reverted or who reverted it. cc @ezyang

**Comment by ZainRizvi** [comment]
I'll look for someone to help you with this @kurtamohler, but the assert that's failing internally looks pretty straighforward:

It's expecting `assert torch.get_default_dtype() == torch.float` to pass, which this PR now violates

**Comment by kurtamohler** [comment]
Thanks @ZainRizvi ! I assume that assert is one of the ones I added in `torch/testing/_internal/common_utils.py` for either `TestCase.setUp` or `TestCase.tearDown`, right? I could just remove those asserts if we need to

**Comment by awgu** [comment]
@kurtamohler Some RPC tests are failing internally from the assertion, but unfortunately, no one is maintaining RPC anymore. I think your suggestion sounds good!

> Or potentially, we could keep the asserts, conditionally disable them by default with a flag, and the tests in pytorch/test could set the flag to enable the asserts. Meta-internal tests would not set the flag, so the asserts would be avoided by default

**Comment by kurtamohler** [comment]
Sounds good, I will submit a new PR shortly

**Comment by kurtamohler** [comment]
After some thinking, I'm not sure how I would enable the check for all the tests in `pytorch/test` and disable for the Meta-internal tests. I wouldn't want to add the setting to every single `if __name__ == '__main__':` in `pytorch/test`, because there are several hundreds of those. 

An idea I had was to check the traceback in `TestCase.__init__` to see if the first frame is a file in the test directory, but that seems very dubious.

So I think I'll only add the setting to the test files which contain tests that temporarily change the default dtype

**Comment by huydhn** [comment]
@pytorchbot drci

(Please ignore this, I'm testing Dr. CI)
